lit.site.cfg

```diff
@@ -0,0 +1,195 @@
+# -*- Python -*-
```

Comment:
I get this is here in order to test the library in the same way that Dafny itself is tested, but it's a poor match for an actual Dafny code base since it's only testing that everything verifies, and doesn't need to assert what happens on invalid Dafny files as the Dafny test suite does.

I'd be happier to see at least a few `{:test}` methods that CI can run instead, as demonstrated by this test file: https://github.com/dafny-lang/dafny/blob/master/Test/DafnyTests/TestAttribute.dfy. It's a little unfortunate that the only supported way to run them is via `dotnet test`, but as I just suggested on https://github.com/dafny-lang/dafny/issues/1322#issuecomment-889184653 perhaps we can add a `dafny test` command to make that simpler.

Yes, I know, it probably feels weird that I'm asking for dynamically executed tests on a Dafny library. In my experience it helps to guard against specification errors, though.

Reply:
Yes, we can probably slim down the `lit.site.cfg` file.  As you say, making sure everything continues to verify seems like the most critical aspect.

Reply:
Also, what if a library function has contradictory preconditions and thus is unusable? Or if its postcondition is so weak that nothing useful can be proven with it? IMHO, there should also be basic usability tests that run on CI, in the form of usage examples. That's why I added the following sentence to the README when I started this repo: 

> All added code should be accompanied by some examples in the `examples` directory which illustrates how to use the code.

It might need some work to add examples, but I think it would be well worth doing, because examples that are run on CI are a great form of never-outdated documentation.

Reply:
I would drop lit entirely, personally, and just rely on a single call to `dafny` with all source files to compile them to C# (which will ensure they validate), and then run the resulting xUnit tests.

It's probably a good idea to have a separate `tests` directory from `examples` since they serve different purposes. A common pattern I've enforced before is that `tests` is where all the tests go, but each entry in `examples` always gets a simple test that executes it as well.

Reply:
@robin-aws I'm not sure I understand your suggestion.  We still want to make sure the library files verify, which compilation won't help with.  I'm fine with using something other than `lit` for that if you have something concrete in mind.

@samuelgruetter I think one of the benefits of verification is that a method/lemma signature already provides detailed, machine-checkable documentation of the expectations of the inputs/outputs.  If there's something surprising about how a particular method/lemma works, an example could be useful, but I think for most of what we're adding here, the examples would be trivial and uninteresting.  Contributions are welcome, however :)  

> what if a library function has contradictory preconditions and thus is unusable? Or if its postcondition is so weak that nothing useful can be proven with it? 

Yes, that would be worth looking for when reviewing a proposed addition to the library.  Also something users will presumably discover quite quickly if we miss something.  I suspect the kinds of subtle problems that will escape inspection would also fail to be exercised by a trivial example as well.

Reply:
Sorry, I could have been clearer with my suggestion. Besides what @samuelgruetter is saying about testing that pre- and post-conditions make sense, I want the CI to ensure the code can be compiled and run and produce the right results. Given that there are lots of `function method`s in this library and not just `function`s, it should be a requirement to ensure that the code is actually compilable and works correctly. 

Here's a really simple example:

```dafny
module SeqTest {
  import Seq
  method {:test} TestApply() {
    var input := [1, 2, 3, 4, 5];
    var output := Seq.apply(i => i * 2, input);
    expect output == [2, 4, 6, 8, 10];
  }
}
```

Building and testing would be a two step process:

```
dafny /compile:0 /spillTargetCode:2 <all Dafny source files>
```

That will still run verification on all the source files and potentially fail. If verification succeeds, you would then execute the resulting C# tests:

```
dotnet test
```

This would require a `.csproj` file similar to the one used in the Dafny test suite: https://github.com/dafny-lang/dafny/blob/master/Test/DafnyTests/DafnyTests.csproj.

The `{:test}` attribute is only handled by the C# compiler for now but we're adding support for the other target languages. Plus, as I mentioned in https://github.com/dafny-lang/dafny/issues/1322#issuecomment-889184653, we should make the process a single simple step by adding something like a `dafny test` command.